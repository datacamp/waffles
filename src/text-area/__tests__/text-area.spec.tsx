import { useRef, useEffect } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { TextArea } from '../index';

const MOCKED_ID = '123abC';

jest.mock('nanoid', () => {
  return {
    nanoid: () => MOCKED_ID,
  };
});

function TestRefTextArea() {
  const inputRef = useRef<HTMLTextAreaElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  return <TextArea ref={inputRef} label="How do you feel?" />;
}

describe('TextArea', () => {
  it('renders a basic text area', () => {
    const { container } = render(<TextArea label="How do you feel?" />);

    const textarea = container.querySelector('textarea');
    const label = container.querySelector('label');

    expect(textarea).toBeInTheDocument();
    expect(label).toHaveTextContent('How do you feel?');
  });

  it('renders a text area with description and placeholder', () => {
    const { getByPlaceholderText, getByText } = render(
      <TextArea
        label="How do you feel?"
        description="Additional description"
        placeholder="Leave a comment here"
      />,
    );

    const textarea = getByPlaceholderText('Leave a comment here');
    const description = getByText('Additional description');

    expect(textarea).toBeInTheDocument();
    expect(description).toBeInTheDocument();
  });

  it('handles typing correctly', () => {
    const handleChange = jest.fn();
    const { getByPlaceholderText } = render(
      <TextArea
        onChange={handleChange}
        label="How do you feel?"
        placeholder="Leave a comment here"
      />,
    );

    const textarea = getByPlaceholderText('Leave a comment here');
    userEvent.type(textarea, 'test');

    expect(handleChange).toHaveBeenCalledTimes(4);
  });

  it('after clicking on label text area got focused', () => {
    const handleFocus = jest.fn();
    const { getByText } = render(
      <TextArea
        label="How do you feel?"
        description="Additional description"
        onFocus={handleFocus}
      />,
    );

    const label = getByText('How do you feel?');
    userEvent.click(label);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('renders additional message when text area is required', () => {
    const { getByText } = render(
      <TextArea label="How do you feel?" required />,
    );

    const message = getByText(/required/i);

    expect(message).toBeInTheDocument();
  });

  it('renders error message', () => {
    const { getByText } = render(
      <TextArea label="How do you feel?" error="Comment is just too long." />,
    );

    const error = getByText('Comment is just too long.');

    expect(error).toBeInTheDocument();
  });

  it('sets correct aria attributes when has error', () => {
    const { getByPlaceholderText } = render(
      <TextArea
        label="How do you feel?"
        error="Comment is just too long."
        placeholder="Leave a comment here"
      />,
    );

    const textarea = getByPlaceholderText('Leave a comment here');

    expect(textarea).toHaveAttribute('aria-invalid', 'true');
  });

  it('textarea and label are associated by the same ID', () => {
    const { container } = render(<TextArea label="How do you feel?" />);

    const textarea = container.querySelector('textarea');
    const label = container.querySelector('label');

    expect(label).toHaveAttribute('for', `form-control-${MOCKED_ID}`);
    expect(textarea).toHaveAttribute('id', `form-control-${MOCKED_ID}`);
  });

  it('when user defined ID is passed, textarea and label are associated by it', () => {
    const { container } = render(
      <TextArea label="How do you feel?" id="mood-textarea" />,
    );

    const textarea = container.querySelector('textarea');
    const label = container.querySelector('label');

    expect(label).toHaveAttribute('for', `mood-textarea`);
    expect(textarea).toHaveAttribute('id', `mood-textarea`);
  });

  it('textarea and error are associated by the same ID', () => {
    const { container, getByText } = render(
      <TextArea label="How do you feel?" error="Comment is just too long." />,
    );

    const textarea = container.querySelector('textarea');
    const error = getByText('Comment is just too long.');

    expect(textarea).toHaveAttribute(
      'aria-errormessage',
      `form-control-error-${MOCKED_ID}`,
    );
    expect(error).toHaveAttribute('id', `form-control-error-${MOCKED_ID}`);
  });

  it('sets the data attribute on the text area', () => {
    const { getByTestId } = render(
      <TextArea data-testid="mood-textarea" label="How do you feel?" />,
    );

    const input = getByTestId('mood-textarea');

    expect(input).toBeInTheDocument();
  });

  it('renders the disabled text area', () => {
    const { container } = render(
      <TextArea disabled label="How do you feel?" />,
    );

    const textarea = container.querySelector('textarea');

    expect(textarea).toBeDisabled();
  });

  it('accepts ref and could be focused programmatically', () => {
    const { container } = render(<TestRefTextArea />);

    const textarea = container.querySelector('textarea');

    expect(textarea).toHaveFocus();
  });

  it('renders snapshot with label and description', () => {
    const { container } = render(
      <TextArea
        label="How do you feel?"
        description="Additional description"
        placeholder="Leave a comment here"
      />,
    );

    const textarea = container.firstChild;
    expect(textarea).toMatchSnapshot();
  });

  it('renders snapshot of required with error', () => {
    const { container } = render(
      <TextArea
        label="How do you feel?"
        error="Comment is just too long."
        required
      />,
    );

    const textarea = container.firstChild;
    expect(textarea).toMatchSnapshot();
  });

  it('renders snapshot of inverted', () => {
    const { container } = render(
      <TextArea
        inverted
        label="How do you feel?"
        description="Additional description"
      />,
    );

    const textarea = container.firstChild;
    expect(textarea).toMatchSnapshot();
  });

  it('renders snapshot of disabled', () => {
    const { container } = render(
      <TextArea
        disabled
        label="How do you feel?"
        description="Additional description"
      />,
    );

    const textarea = container.firstChild;
    expect(textarea).toMatchSnapshot();
  });

  it('renders all supporting elements to make auto-grow work', () => {
    const { container } = render(
      <TextArea autoGrow label="How do you feel?" />,
    );

    const textarea = container.firstChild;
    expect(textarea).toMatchSnapshot();
  });
});
