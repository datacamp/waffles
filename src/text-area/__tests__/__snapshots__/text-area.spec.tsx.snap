// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders all supporting elements to make auto-grow work 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-weight: 800;
  color: #05192D;
  margin-bottom: 8px;
}

.emotion-2 {
  display: grid;
}

.emotion-3 {
  position: relative;
  grid-area: 1/1/2/2;
}

.emotion-3::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-4 {
  display: block;
  width: 100%;
  height: 100%;
  resize: none;
  padding: 8px 12px;
  color: #05192D;
  background-color: #FFFFFF;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  border: 1px solid #848C92;
  border-radius: 4px;
  outline: 0;
  color: #05192D;
  background-color: #FFFFFF;
}

.emotion-4:disabled {
  cursor: default;
  opacity: 0.6;
}

.emotion-4::-webkit-input-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4::-moz-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4:-ms-input-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4::placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-5 {
  display: block;
  width: 100%;
  height: 100%;
  resize: none;
  padding: 8px 12px;
  color: #05192D;
  background-color: #FFFFFF;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  border: 1px solid #848C92;
  border-radius: 4px;
  outline: 0;
  grid-area: 1/1/2/2;
  visibility: hidden;
  white-space: pre-wrap;
}

.emotion-5:disabled {
  cursor: default;
  opacity: 0.6;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-control-123abC"
  >
    How do you feel?
  </label>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-3"
    >
      <textarea
        class="emotion-4"
        id="form-control-123abC"
      />
    </div>
    <div
      aria-hidden="true"
      class="emotion-5"
    >
      undefined 
    </div>
  </div>
</div>
`;

exports[`TextArea renders snapshot of disabled 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-weight: 800;
  color: #05192D;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.5;
  margin-bottom: 8px;
  color: #05192D;
  font-size: 14px;
  line-height: 1.25;
  margin-top: -4px;
}

.emotion-3 {
  position: relative;
}

.emotion-3::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-4 {
  display: block;
  width: 100%;
  height: 100%;
  resize: none;
  padding: 8px 12px;
  color: #05192D;
  background-color: #FFFFFF;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  border: 1px solid #848C92;
  border-radius: 4px;
  outline: 0;
  color: #05192D;
  background-color: #FFFFFF;
}

.emotion-4:disabled {
  cursor: default;
  opacity: 0.6;
}

.emotion-4::-webkit-input-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4::-moz-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4:-ms-input-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4::placeholder {
  opacity: 1;
  color: #65707C;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-control-123abC"
  >
    How do you feel?
  </label>
  <p
    class="emotion-2"
  >
    Additional description
  </p>
  <div
    class="emotion-3"
  >
    <textarea
      class="emotion-4"
      disabled=""
      id="form-control-123abC"
    />
  </div>
</div>
`;

exports[`TextArea renders snapshot of inverted 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-weight: 800;
  color: #FFFFFF;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.5;
  margin-bottom: 8px;
  color: #FFFFFF;
  font-size: 14px;
  line-height: 1.25;
  margin-top: -4px;
}

.emotion-3 {
  position: relative;
}

.emotion-3::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-4 {
  display: block;
  width: 100%;
  height: 100%;
  resize: none;
  padding: 8px 12px;
  color: #05192D;
  background-color: #FFFFFF;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  border: 1px solid #848C92;
  border-radius: 4px;
  outline: 0;
  color: #FFFFFF;
  background-color: #213147;
}

.emotion-4:disabled {
  cursor: default;
  opacity: 0.6;
}

.emotion-4::-webkit-input-placeholder {
  opacity: 1;
  color: #9BA3AB;
}

.emotion-4::-moz-placeholder {
  opacity: 1;
  color: #9BA3AB;
}

.emotion-4:-ms-input-placeholder {
  opacity: 1;
  color: #9BA3AB;
}

.emotion-4::placeholder {
  opacity: 1;
  color: #9BA3AB;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-control-123abC"
  >
    How do you feel?
  </label>
  <p
    class="emotion-2"
  >
    Additional description
  </p>
  <div
    class="emotion-3"
  >
    <textarea
      class="emotion-4"
      id="form-control-123abC"
    />
  </div>
</div>
`;

exports[`TextArea renders snapshot of required with error 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-weight: 800;
  color: #05192D;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  font-size: 12px;
  color: #65707C;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-3 {
  font-size: 14px;
  color: #DD3400;
}

.emotion-4 {
  position: relative;
}

.emotion-4::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-5 {
  display: block;
  width: 100%;
  height: 100%;
  resize: none;
  padding: 8px 12px;
  color: #05192D;
  background-color: #FFFFFF;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  border: 1px solid #848C92;
  border-radius: 4px;
  outline: 0;
  color: #05192D;
  background-color: #FFFFFF;
  border-color: #DD3400;
  box-shadow: 0 0 0 1px #DD3400;
}

.emotion-5:disabled {
  cursor: default;
  opacity: 0.6;
}

.emotion-5::-webkit-input-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-5::-moz-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-5:-ms-input-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-5::placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-6 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  position: absolute;
  z-index: 1;
  bottom: -8px;
  left: 0;
  -webkit-transform: translateY(100%);
  -moz-transform: translateY(100%);
  -ms-transform: translateY(100%);
  transform: translateY(100%);
  color: #C01100;
  line-height: 1.25;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-control-123abC"
  >
    How do you feel?
    <span
      aria-hidden="true"
      class="emotion-2"
    >
      <span
        class="emotion-3"
      >
        â€¢
      </span>
       Required
    </span>
  </label>
  <div
    class="emotion-4"
  >
    <textarea
      aria-describedby="form-control-error-123abC"
      aria-errormessage="form-control-error-123abC"
      aria-invalid="true"
      class="emotion-5"
      id="form-control-123abC"
    />
  </div>
  <p
    class="emotion-6"
    id="form-control-error-123abC"
  >
    Comment is just too long.
  </p>
</div>
`;

exports[`TextArea renders snapshot with label and description 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25;
  font-weight: 800;
  color: #05192D;
  margin-bottom: 8px;
}

.emotion-2 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.5;
  margin-bottom: 8px;
  color: #05192D;
  font-size: 14px;
  line-height: 1.25;
  margin-top: -4px;
}

.emotion-3 {
  position: relative;
}

.emotion-3::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 6px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-4 {
  display: block;
  width: 100%;
  height: 100%;
  resize: none;
  padding: 8px 12px;
  color: #05192D;
  background-color: #FFFFFF;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  border: 1px solid #848C92;
  border-radius: 4px;
  outline: 0;
  color: #05192D;
  background-color: #FFFFFF;
}

.emotion-4:disabled {
  cursor: default;
  opacity: 0.6;
}

.emotion-4::-webkit-input-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4::-moz-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4:-ms-input-placeholder {
  opacity: 1;
  color: #65707C;
}

.emotion-4::placeholder {
  opacity: 1;
  color: #65707C;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="form-control-123abC"
  >
    How do you feel?
  </label>
  <p
    class="emotion-2"
  >
    Additional description
  </p>
  <div
    class="emotion-3"
  >
    <textarea
      class="emotion-4"
      id="form-control-123abC"
      placeholder="Leave a comment here"
    />
  </div>
</div>
`;
